Added power binop

royels opened the pull request:
Addresses issue #1629 
@brian-brazil I thought I’d have a go at it, no pun intended. I used the `^` syntax.

fabxc commented:
Thanks a lot!

Did you use `^` because you think it's more understandable or because having two character binops require some larger tweaks to be implemented?
Either is perfectly fine – just want to make sure we pick the most intuitive operator here.

royels commented:
Both! I actually tried using `**` first because I’m a Python person, but from the errors that came up I realized it was interpreting that as two consecutive `*` ’s despite my having added the two character binop to the lexer.

So I ended up changing it to `^`, which is also fairly intuitive. And it’s single character, so I didn’t get the same lexing errors as `**`.

brian-brazil commented:
The thing to consider is if we'll ever have bitwise operators, in particular a bitwise xor as that's `^` in C. I think it's fairly unlikely we'll end up with such operators, and we can always give them more descriptive names.

brian-brazil commented:
Could you add a test in `operators.test`?

fabxc commented:
We essentially have to consider whether we take a mathematics or programming approach.
I think generally doing the mathematical things works well and we might just use `xor` (if ever) for clarities sake.

royels commented:
@brian-brazil I can’t seem to find operators.test, even after grepping. Is that a different name for something? I added a test to the test operators section of lex_test, if that’s the same thing.

brian-brazil commented:
https://github.com/prometheus/prometheus/blob/master/promql/testdata/operators.test

fabxc commented:
@royels ping :)

royels commented:
Added a couple of tests for the power binop to operators.test!

fabxc commented:
I just realized we have not specified the operator precedence for the power operator.
Currently it falls through to "lowest precedence" in [this function](https://github.com/royels/prometheus/blob/625eb30591e7be09b20a7c46b23ec7aed686feed/promql/lex.go#L88-L106).

In Ruby and Python for example the precedence is quite high.
I would argue for making it of precedence level `6` and add tests to verify that – unless someone feels different (@brian-brazil?)

Also, could you squash the commits and make it `promql: add power binary operation`?

beorn7 commented:
Also note that the power operator is usually expected to be right-associative, i.e. `a ^ b ^ c == a ^ (b ^ c)`.

fabxc commented:
Good catch, that will require quite a few more tweaks.

brian-brazil commented:
6 right-associative sounds right to me.

royels commented:
It works!

Sorry for the absence again, between finals and work starting it was…hectic.
After reading code for quite a bit of time, I *think* this is the needed change.

That being said, the fact that I specified it to `itemPOW` (the `^` operator) makes it feel a little hardcoded. Is that something I should change?

fabxc commented on promql/parse.go:
```diff
@@ -500,7 +500,8 @@ func (p *parser) expr() Expr {
 func (p *parser) balance(lhs Expr, op itemType, rhs Expr, vecMatching *VectorMatching, returnBool bool) *BinaryExpr {
-	if lhsBE, ok := lhs.(*BinaryExpr); ok && lhsBE.Op.precedence() < op.precedence() {
+	if lhsBE, ok := lhs.(*BinaryExpr); (ok && lhsBE.Op.precedence() < op.precedence()) ||
+		(ok && lhsBE.Op.precedence() == op.precedence() && lhsBE.Op == itemPOW && op == itemPOW) {
```
Would indeed be good to encapsulate associativity logic here for future clarity. We could define another method similar to `precedence()`; `isRightAssociative()` seems simple enough.

To make the code more self-explanatory we could structure this condition something like:

```go
lhsBE, ok := lhs.(*BinaryExpr)
precd := lhsBe.Op.precedence() - op.precedence()
// Relies on the invariant that no operators with the same precedence level can have different associativity.
if ok && (precd < 0 || precd == 0 && op.isRightAssociative()) {
    // ...
}
```

Does that make sense?

fabxc commented on promql/testdata/operators.test:
```diff
@@ -35,6 +34,26 @@ eval instant at 50m SUM(http_requests) BY (job) % 3
+	{job="api-server"} 488
+	{job="app-server"} 40
+
+eval instant at 50m SUM(http_requests) BY (job) % 2 ^ 3 ^ 2 ^ 2
```
Can we add a test where we are doing `^` between two vectors?

fabxc commented:
Thanks for working on this – especially in stressful times!
Resolving the hard coded bit would indeed be good.

royels commented:
Done! (Part Deuce)
Changed hardcode to a switch statement in `isRightAssociative`, so future right-assoc. operators can be added cleanly. 
Added the test between two vectors as well!

fabxc commented:
Great! 👍
